common/src/java/com/zimbra/common/soap/W3cDomUtil.java

```diff
@@ -91,29 +80,58 @@ private W3cDomUtil() {}
             }
     };
 
+    public static DocumentBuilderFactory makeDocumentBuilderFactory() throws ParserConfigurationException {
+        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+        dbf.setNamespaceAware(true);
+        dbf.setIgnoringComments(true);
+        // protect against recursive entity expansion DOS attack and perhaps other things
+        dbf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+        try {
+            // XXE attack prevention
+            dbf.setFeature(Constants.DISALLOW_DOCTYPE_DECL, true);
+            dbf.setFeature(Constants.EXTERNAL_GENERAL_ENTITIES, false);
+            dbf.setFeature(Constants.EXTERNAL_PARAMETER_ENTITIES, false);
+            dbf.setFeature(Constants.LOAD_EXTERNAL_DTD, false);
+            dbf.setXIncludeAware(false);
+            dbf.setExpandEntityReferences(false);
+        } catch (IllegalArgumentException iae) {
+            ZimbraLog.misc.debug("Disabling doctype-decl not supported", iae);
```

Comment:
Please have a look at https://github.com/Zimbra/zm-mailbox/pull/242

1) IllegalArgumentException is removed
2) ParserConfigurationException is already added as throws, which is the only exception thrown by setFeature() method.
3) Reply to point 3 - Googled around the order and try-catching separately but i did not see try-catching separately anywhere.